src/anndata/_core/xarray.py

```diff
@@ -195,4 +275,79 @@ def __setitem__(self, key, value):
         else:
             # maintain setting behavior of a 2D dataframe i.e., one dim
             value = (self.index_dim, value)
-        super().__setitem__(key, value)
+        self.ds.__setitem__(key, value)
+
+    def copy(self, *args, **kwargs):
+        """
+        Return a copy of the Dataset2D object.
+        """
+        as_2d = Dataset2D(self.ds.copy(*args, **kwargs))
+        as_2d.true_index_dim = self.true_index_dim
+        as_2d.is_backed = self.is_backed
+        return as_2d
+
+    def isel(self, *args, **kwargs):
+        """
+        Return a isel of the Dataset2D object.
+        """
+        as_2d = Dataset2D(self.ds.isel(*args, **kwargs))
+        as_2d.true_index_dim = self.true_index_dim
+        as_2d.is_backed = self.is_backed
+        return as_2d
+
+    def __iter__(self) -> Iterator[Hashable]:
+        return iter(self.ds)
+
+    def __len__(self) -> int:
+        return len(self.ds)
+
+    @property
+    def dtypes(self) -> pd.Series:
+        """
+        Return a Series with the dtypes of the variables in the Dataset2D.
+        """
+        return self.ds.dtypes
+
+    def equals(self, b: object) -> bool:
+        if isinstance(b, Dataset2D):
+            b = b.ds
+        return self.ds.equals(b)
+
+    def reindex(
+        self,
+        index: pd.Index | None = None,
+        axis: Literal[0] = 0,
+        fill_value: Any | None = np.nan,
+    ) -> Dataset2D:
+        """Reindex the current object against a new index.
+
+        Parameters
+        ----------
+        index
+            The new index for reindexing, by default None
+        axis
+            Provided for API consistency, should not be called over axis!=0, by default 0
+        fill_value
+            The value with which to fill in via :meth:`pandas.Series.reindex`, by default np.nan
+
+        Returns
+        -------
+            Reindexed dataset.
+        """
+        index_dim = self.index_dim
+        if axis == 0:
+            # Dataset.reindex() can't handle ExtensionArrays
+            extension_arrays = {
+                col: arr.data
+                for col, arr in self.ds.items()
+                if pd.api.types.is_extension_array_dtype(arr.dtype)
+            }
+            el = self.ds.drop_vars(extension_arrays.keys())
+            el = el.reindex({index_dim: index}, method=None, fill_value=fill_value)
+            for col, arr in extension_arrays.items():
+                el[col] = pd.Series(arr, index=self.index).reindex(
+                    index, fill_value=fill_value
+                )
+            return Dataset2D(el)
+        msg = "This should be unreachable, please open an issue."
+        raise Exception(msg)
```

Comment:
I guess if `Dataset2D` is a public API and this is a public API, this should be a `ValueError` and tell users to not do that instead.

Also plain `Exception` is iffy, as originally worded this would be an `AssertionError` I guess. Please start asking me if you're unsure which exception class is appropriate.

could you also make this an early exit and add `# pragma: no cover`?

Reply:
before I forget: `# pragma: no cover` or test coverage?

Reply:
thanks for the reminder!